client/src/controllers/MessagesController.test.js

```diff
@@ -0,0 +1,232 @@
+import { Application } from '@hotwired/stimulus';
```

Comment:
These unit tests do a good job, thank you, however I recommend a few things in general when writing tests

1. Avoid all possible HTML content that does not relate to the test functionality (e.g. Django comments or other aria props, wrapper divs.
2. Make sure HTML is correctly formatted (your snippets do not have a closing div tag) and use simpler spacing (two spaces) and consistent in/out tabs, all of this is to ensure readability.
3. Try to think of describe/it as folders and tasks inside the folder. You are heading in this direction but it just needs a bit of refinement. Maybe two describe blocks, one for the bare bones basic behaviour and other for the more complex behaviour. Inside each describe block, try to break up your assertion things into `it` (e.g. adding no messages, adding first message, adding second message).
4. Try not to be too prescriptive in your selectors, you want a balance between the tests testing the right thing but not being too brittle.
5. Finally, when testing timers, do not try to handle this yourself, use Jest mock timers, it will make the unit tests work correctly and avoid edge case problems. see jest. use fake timers docs

All in all, here is a recommended approach for unit testing here.

```js
import { Application } from '@hotwired/stimulus';
import { MessagesController } from './MessagesController';

jest.useFakeTimers();

describe('MessagesController', () => {
  let application;

  afterEach(() => {
    jest.clearAllTimers();
  });

  describe('default behaviour', () => {
    beforeAll(() => {
      application?.stop();
      document.body.innerHTML = `
    <div
      class="messages"
      data-controller="w-messages"
      data-action="w-messages:add@document->w-messages#add"
    >
      <ul data-w-messages-target="container"></ul>
      <template data-w-messages-target="templates">
        <li class="success" data-message-status="success"><span></span></li>
      </template>
    </div>`;

      application = Application.start();
      application.register('w-messages', MessagesController);
    });

    it('should not add elements when connected by default', () => {
      expect(document.querySelectorAll('li')).toHaveLength(0);
    });

    it('should allow for a message to be added via the add method', () => {
      const text = 'first message text';

      document.dispatchEvent(
        new CustomEvent('w-messages:add', { detail: { text } }),
      );

      // the item should be added
      const item = document.querySelector('li');

      expect(item.classList.toString()).toEqual('success');
      expect(item.lastElementChild.textContent).toEqual(text);
    });

    it('should allow for a second message to be added', async () => {
      expect(document.querySelectorAll('li')).toHaveLength(1);

      const text = 'second message text';

      document.dispatchEvent(
        new CustomEvent('w-messages:add', { detail: { text } }),
      );

      expect(document.querySelectorAll('li')).toHaveLength(2);

      // the item should be added
      const item = document.querySelector('li:last-child');

      expect(item.classList.toString()).toEqual('success');
      expect(item.lastElementChild.textContent).toEqual(text);
    });
  });

  describe('additional behaviour', () => {
    beforeAll(() => {
      application?.stop();
      document.body.innerHTML = `
    <div
      class="messages"
      data-controller="w-messages"
      data-action="w-messages:add@document->w-messages#add"
      data-w-messages-added-class="new"
      data-w-messages-show-class="appear"
    >
      <ul data-w-messages-target="container"></ul>
      <template data-w-messages-target="templates">
        <li class="success" data-message-status="success"><span></span></li>
        <li class="error" data-message-status="error"><strong></strong></li>
        <li class="warning" data-message-status="warning"><span></span></li>
      </template>
    </div>`;

      application = Application.start();
      application.register('w-messages', MessagesController);
    });

    it('should not add any classes when connected by default', () => {
      expect(document.querySelector('.messages').classList.toString()).toEqual(
        'messages',
      );

      expect(document.querySelectorAll('li')).toHaveLength(0);
    });

    it('should allow for a message to be added via the add method', () => {
      const text = 'first message text';

      document.dispatchEvent(
        new CustomEvent('w-messages:add', { detail: { text } }),
      );

      // set the new class on the container
      expect(
        document.querySelector('.messages').classList.contains('new'),
      ).toBe(true);

      // the item should be added
      const item = document.querySelector('li');

      expect(item.classList.toString()).toEqual('success');
      expect(item.lastElementChild.textContent).toEqual(text);

      // it should add a shown class to the message after the timeout
      jest.runAllTimers();

      expect(document.querySelector('.messages').classList.toString()).toEqual(
        'messages new appear',
      );
    });

    it('should allow for a second message to be added with a specific status', async () => {
      expect(document.querySelectorAll('li')).toHaveLength(1);

      const text = 'second message text';

      document.dispatchEvent(
        new CustomEvent('w-messages:add', {
          detail: { text, status: 'warning' },
        }),
      );

      expect(document.querySelectorAll('li')).toHaveLength(2);

      // the item should be added
      const item = document.querySelector('li:last-child');

      expect(item.classList.toString()).toEqual('warning');
      expect(item.lastElementChild.textContent).toEqual(text);
    });

    it('should allow for any last child in the matched template to have content replaced', async () => {
      expect(document.querySelectorAll('li')).toHaveLength(2);

      const text = 'third message text';

      document.dispatchEvent(
        new CustomEvent('w-messages:add', {
          detail: { text, status: 'error' },
        }),
      );

      expect(document.querySelectorAll('li')).toHaveLength(3);

      // the item should be added
      const item = document.querySelector('li:last-child');

      expect(item.classList.toString()).toEqual('error');
      // note: finding the strong element
      expect(item.querySelector('strong').textContent).toEqual(text);
    });

    it('should allow for items to be cleared when adding a new one', () => {
      expect(document.querySelectorAll('li')).toHaveLength(3);

      const text = 'new message text';

      document.dispatchEvent(
        new CustomEvent('w-messages:add', {
          detail: { clear: true, text, status: 'warning' },
        }),
      );

      expect(document.querySelectorAll('li')).toHaveLength(1);

      const item = document.querySelector('li');

      expect(item.classList.toString()).toEqual('warning');

      expect(item.lastElementChild.textContent).toEqual(text);
    });
  });
});
```

Reply:
<img width="859" alt="Screenshot 2023-03-08 at 8 45 44 am" src="https://user-images.githubusercontent.com/1396140/223571246-15ebe6d0-82d1-490d-a14a-f29087b52504.png">

If you look at the jest ouput you can see this makes for a nice readable log of what's being tested and what 'groups' it is in.